src/overload_gradient.jl

```diff
@@ -1,67 +1,144 @@
-for fn in union(ops_1_to_1_s, ops_1_to_1_f)
-    @eval Base.$fn(t::GlobalGradientTracer) = t
+## 1-to-1
+for fn in ops_1_to_1
+    @eval function Base.$fn(t::T) where {T<:GradientTracer}
+        if is_firstder_zero_global($fn)
+            return empty(T)
+        else
+            return t
+        end
+    end
+    @eval function Base.$fn(t::D) where {P,T<:GradientTracer,D<:Dual{P,T}}
+        x = primal(t)
+        out = Base.$fn(x)
+        if is_firstder_zero_local($fn, x)
+            return Dual(out, empty(T))
+        else
+            return Dual(out, tracer(t))
+        end
+    end
 end
 
-for fn in union(ops_1_to_1_z, ops_1_to_1_const)
-    @eval Base.$fn(::T) where {T<:GlobalGradientTracer} = empty(T)
-end
+## 2-to-1
+for fn in ops_2_to_1
+    @eval function Base.$fn(tx::T, ty::T) where {T<:GradientTracer}
+        ∂f∂x0 = is_firstder_arg1_zero_global($fn)
+        ∂f∂y0 = is_firstder_arg2_zero_global($fn)
+        if ∂f∂x0
+            if ∂f∂y0
+                return empty(T)
+            else # ∂f∂y ≠ 0 
+                return ty
+            end
+        else # ∂f∂x ≠ 0 
+            if ∂f∂y0
+                return tx
+            else # ∂f∂y ≠ 0 
+                return T(gradient(tx) ∪ gradient(ty))
+            end
+        end
```

Comment:
how about this (schematically) to avoid the nesting?
```julia
left = ∂f∂x0 ? empty(T) : tx
right = ∂f∂y0 ? empty(T) : ty
return left ∪ right
```

Reply:
I'll have a look at which type of control flow statement works best for constant propagation and dead code elimination.

Reply:
My generic take on this is that any kind of compiler optimization is peanuts compared to the cost of the operations we are doing, which is merging sets. So I would bed on code clarity first and foremost